### PR TITLE
Support automatic derivation of Trials<X> for structured classes in Java and Scala

### LINKED ISSUE #6: Support automatic derivation of `Trials<X>` for a structured X in Java.
On the Scala side, there is already support for automatic derivation of `Trials[X]` where `X` is some algebraic data type that heads a hierarchy of sum and product types - courtesy of [Magnolia](https://github.com/propensive/magnolia) - nice job there.

What we lack is the same experience on the Java side - and for that matter for classes in Scala that are not sum / product types, ie. 'conventional OO classes in Scala'.

One approach would be to use the public constructors of `X` to guide the assembly of a `Trials<X>`, using a mixture of reflection and ByteBuddy for the implementation.

Ideally this should work for both Java and Scala code, although Java is the priority here.

### PR DESCRIPTION
**Pull request by sageserpent-open**
Added automatic derivation of `Trials<X>` for structured Java classes, Java records, enums, containers, arrays, interfaces/abstract classes, and conventional OO Scala classes via `Trials.api().instances(Class<X>)` and `instances(Type)`. Implemented using public constructor reflection, ByteBuddy bytecode generation/invokers, and Caffeine caching.

Fixes #6

---
*PR created automatically by Jules for task [8245826892877412278](https://jules.google.com/task/8245826892877412278) started by @sageserpent-open*

### CODE REVIEW

File: core-library/src/test/java/com/sageserpent/americium/java/InstancesJavaTest.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        api.instances(Instant.class).withLimit(10).supplyTo(val -> assertThat(val, is(notNullValue())));
+        api.instances(BigInteger.class).withLimit(10).supplyTo(val -> assertThat(val, is(notNullValue())));
+        api.instances(BigDecimal.class).withLimit(10).supplyTo(val -> assertThat(val, is(notNullValue())));
+    }
+
+    @Test
+    void testEnumDerivation() {
+        Set<Season> seasonsSeen = new HashSet<>();
+        api.instances(Season.class).withLimit(20).supplyTo(seasonsSeen::add);
+        assertThat(seasonsSeen.size(), greaterThanOrEqualTo(1));
```
Comment (sageserpent-open) [inline core-library/src/test/java/com/sageserpent/americium/java/InstancesJavaTest.java:105]: Can we assert that we see `SeasonsSeen.values().length` or whatever here (can't remember how to enumerate the constants in Java!).

File: core-library/src/test/java/com/sageserpent/americium/java/InstancesJavaTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    void testRecordDerivation() {
+        api.instances(Point.class).withLimit(20).supplyTo(point -> {
+            assertThat(point.x(), is(notNullValue()));
+            assertThat(point.y(), is(notNullValue()));
+        });
+    }
+
+    @Test
+    void testGenericClassDerivation() {
+        Trials<Box<String>> stringBoxTrials = api.instances(new com.google.common.reflect.TypeToken<Box<String>>() {}.getType());
```
Comment (sageserpent-open) [inline core-library/src/test/java/com/sageserpent/americium/java/InstancesJavaTest.java:136]: I didn't know about `TypeToken`, that's useful, thank you.

File: core-library/src/test/java/com/sageserpent/americium/java/InstancesJavaTest.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        stringBoxTrials.withLimit(20).supplyTo(box -> {
+            assertThat(box, is(notNullValue()));
+            assertThat(box.content(), is(notNullValue()));
+        });
+    }
+
+    @Test
+    void testRecursiveClassDerivation() {
+        api.instances(BinaryTree.class).withLimit(30).supplyTo(tree -> {
+            assertThat(tree, is(notNullValue()));
```
Comment (sageserpent-open) [inline core-library/src/test/java/com/sageserpent/americium/java/InstancesJavaTest.java:146]: Could we add a `depth` and a `size` method to `BinaryTree` and bolster the assertion checking to at least say that we get more than one depth and size?

File: core-library/src/test/scala/com/sageserpent/americium/InstancesScalaTest.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package com.sageserpent.americium
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class ConventionalScalaPerson(val name: String, val age: Int)
+
+class ConventionalScalaTree(
```
Comment (sageserpent-open) [inline core-library/src/test/scala/com/sageserpent/americium/InstancesScalaTest.scala:8]: What you've come up with here is valid, but I'm not sure we want to advertise this functionality.

Scala code tends to avoid the use of null references, so yes, you can and will see conventional object-oriented style classes (as opposed to functional style case classs), either imperative or pure, but you won't usually see them with nullable fields unless as some heavily optimised library code - usually it's `Option` that steps in to represent presence versus absense, or some custom union type, or maybe `Either`.

This sends a signal to newcomers to Scala that using null references is the done thing; I'd rather avoid that. Java - fine, Scala - no thank you.

That said, if we made the fields into `Option` than that would be an interesting thing to test.

Comment (sageserpent-open) [inline core-library/src/test/scala/com/sageserpent/americium/InstancesScalaTest.scala:8]: This also raises a question - looking at the Java tests, some of them assert that field values should never be null and others look the other way (`BinaryTree`). Is there some criterion that the implementation uses to decide when to allow nullity, eg a recursive type reference? If so, what happens when two classes are mutually recursive - does the derivation keep track of this?

**Review by sageserpent-open — COMMENTED** [review]
This looks promising, but raises a question about handling of structurally recursive types and nullity.
